src/gpio.rs

```diff
@@ -0,0 +1,447 @@
+    /// Pin number
+    fn pin_id(&self) -> u8;
+    /// Port number starting from 0
+    fn port_id(&self) -> u8;
+}
+
+/// Some alternate mode (type state)
+pub struct Alternate<const A: u8, Otype = PushPull>(PhantomData<Otype>);
+
+/// Input mode (type state)
+pub struct Input;
+
+/// Pull setting for an input.
+#[derive(Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Pull {
+    /// Floating
+    None = 0,
+    /// Pulled up
+    Up = 1,
+    /// Pulled down
+    Down = 2,
+}
+
+/// Open drain input or output (type state)
+pub struct OpenDrain;
+
+/// Output mode (type state)
+pub struct Output<MODE = PushPull> {
+    _mode: PhantomData<MODE>,
+}
+
+/// Push pull output (type state)
+pub struct PushPull;
+
+/// Analog mode (type state)
+pub struct Analog;
+
+/// JTAG/SWD mote (type state)
+pub type Debugger = Alternate<0, PushPull>;
+
+mod marker {
+    /// Marker trait that show if `ExtiPin` can be implemented
+    #[allow(dead_code)] // TODO: Remove when EXTI is implemented
+    pub trait Interruptable {}
+    /// Marker trait for readable pin modes
+    pub trait Readable {}
+    /// Marker trait for slew rate configurable pin modes
+    pub trait OutputSpeed {}
+    /// Marker trait for active pin modes
+    pub trait Active {}
+    /// Marker trait for all pin modes except alternate
+    #[allow(dead_code)] // TODO: Remove when alternate function conversion is implemented
+    pub trait NotAlt {}
+    /// Marker trait for pins with alternate function `A` mapping
+    #[allow(dead_code)] // TODO: Remove when alternate function conversion is implemented
+    pub trait IntoAf<const A: u8> {}
+}
+
+impl<MODE> marker::Interruptable for Output<MODE> {}
+impl marker::Interruptable for Input {}
+impl marker::Readable for Input {}
+impl<const A: u8, MODE> marker::Readable for Alternate<A, MODE> {}
+impl marker::Readable for Output<OpenDrain> {}
+impl marker::Active for Input {}
+impl<Otype> marker::OutputSpeed for Output<Otype> {}
+impl<const A: u8, Otype> marker::OutputSpeed for Alternate<A, Otype> {}
+impl<Otype> marker::Active for Output<Otype> {}
+impl<const A: u8, Otype> marker::Active for Alternate<A, Otype> {}
+impl marker::NotAlt for Input {}
+impl<Otype> marker::NotAlt for Output<Otype> {}
+impl marker::NotAlt for Analog {}
+
+/// GPIO Pin speed selection
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Speed {
+    /// Low speed
+    Low = 0,
+    /// Medium speed
+    Medium = 1,
+    /// High speed
+    High = 2,
+    /// Very high speed
+    VeryHigh = 3,
+}
+
+/// GPIO interrupt trigger edge selection
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Edge {
+    /// Rising edge of voltage
+    Rising,
+    /// Falling edge of voltage
+    Falling,
+    /// Rising and falling edge of voltage
+    RisingFalling,
+}
+
+pub type AF0<Otype = PushPull> = Alternate<0, Otype>;
```

Comment:
Nice tidy-up of the macro that was not saving any lines, although the docstrings are lost.

The docstrings looked ok in the docs, but not a big issue
![image](https://github.com/user-attachments/assets/f0de86c9-9572-4d0c-a69e-04346df533b2)

Reply:
Good catch on the docstrings. I added them back. It's a little less neat, but I'm still not convinced that a macro is necessary.